pkg/render/csr.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package render
+
+import (
+	"fmt"
+	"strings"
+
+	operator "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CreateCSRInitContainer creates an init container that can be added to a pod spec in order to create a CSR for its
+// TLS certificates. It uses the provided params and the k8s downward api to be able to specify certificate subject information.
+func CreateCSRInitContainer(
+	installation *operator.InstallationSpec,
+	cm *operator.CertificateManagement,
+	mountName string,
+	commonName string,
+	keyName string,
+	certName string,
+	dnsNames []string,
+	registerApiserver bool) corev1.Container {
+	return corev1.Container{
+		Name:            "key-cert-provisioner",
+		Image:           components.GetReference(components.ComponentCSRInitContainer, "docker.io/", installation.ImagePath), //todo: Set the right registry.
+		ImagePullPolicy: "Always",                                                                                            //todo: delete this line.
+		VolumeMounts: []corev1.VolumeMount{
+			{MountPath: "/secret", Name: mountName, ReadOnly: false},
```

Comment:
:thinking: Why does the secret need to writable ?

Reply:
An init container will be added to the pod. That init container will submit a CSR and then write the results (private key + cert) to this location. The main container will then read from this location.

Reply:
Hm ... Can we make the secret read only and use a different path to write the certificates ?

Reply:
> Hm ... Can we make the secret read only and use a different path to write the certificates ?
The init container and main container are separate and the init container creates a tls cert that the main container needs  so there _must_ be a shared volume that the init container can write to.

I'm not a huge fan of the directory name though, secret is pretty vague.

Reply:
"/secret" is just a mount path name and has no relationship to kubernetes secrets. It is the location where the in-memory volume is mounted. For the initcontainer it needs write permissions, and the main container has read permissions.

Reply:
Lemme change that name then.

Reply:
Sounds good, let's change it.